js/src/admin/components/UserListPage.tsx

```diff
@@ -0,0 +1,378 @@
+import EditUserModal from '../../common/components/EditUserModal';
+import LoadingIndicator from '../../common/components/LoadingIndicator';
+import Button from '../../common/components/Button';
+
+import icon from '../../common/helpers/icon';
+import listItems from '../../common/helpers/listItems';
+
+import type User from '../../common/models/User';
+
+import ItemList from '../../common/utils/ItemList';
+import classList from '../../common/utils/classList';
+import extractText from '../../common/utils/extractText';
+
+import AdminPage from './AdminPage';
+
+type ColumnData = {
+  /**
+   * Column title
+   */
+  name: String;
+  /**
+   * Component(s) to show for this column.
+   */
+  content: (user: User) => JSX.Element;
+};
+
+type ApiPayload = {
+  data: Record<string, unknown>[];
+  included: Record<string, unknown>[];
+  links: {
+    first: string;
+    next?: string;
+  };
+};
+
+type UsersApiResponse = User[] & { payload: ApiPayload };
+
+/**
+ * Admin page which displays a paginated list of all users on the forum.
+ */
+export default class UserListPage extends AdminPage {
+  /**
+   * Number of users to load per page.
+   */
+  private numPerPage: number = 50;
+
+  /**
+   * Current page number. Zero-indexed.
+   */
+  private pageNumber: number = 0;
+
+  /**
+   * Total number of forum users.
+   *
+   * Fetched from the active `AdminApplication` (`app`), with
+   * data provided by `AdminPayload.php`, or `flarum/statistics`
+   * if installed.
+   */
+  readonly userCount: number = app.data.statistics.users.total;
+
+  /**
+   * Get total number of user pages.
+   */
+  private getTotalPageCount(): number {
+    if (this.userCount === -1) return 0;
+
+    return Math.ceil(this.userCount / this.numPerPage);
+  }
+
+  /**
+   * This page's array of users.
+   *
+   * `undefined` when page loads as no data has been fetched.
+   */
+  private pageData: User[] | undefined = undefined;
+
+  /**
+   * Are there more users available?
+   */
+  private moreData: boolean = false;
+
+  private isLoadingPage: boolean = false;
+
+  /**
+   * Component to render.
+   */
+  content() {
+    if (typeof this.pageData === 'undefined') {
+      this.loadPage(0);
+
+      return [
+        <section class="UserListPage-grid UserListPage-grid--loading">
+          <LoadingIndicator />
+        </section>,
+      ];
+    }
+
+    const columns: (ColumnData & { itemName: string })[] = this.columns().toArray();
+
+    return [
+      <p class="UserListPage-totalUsers">{app.translator.trans('core.admin.users.total_users', { count: this.userCount })}</p>,
+      <section
+        class={classList(['UserListPage-grid', this.isLoadingPage ? 'UserListPage-grid--loadingPage' : 'UserListPage-grid--loaded'])}
+        style={`grid-template-columns: repeat(${columns.length}, minmax(max-content, 300px))`}
+      >
```

Comment:
Right now the cells are all siblings under the same element, I would much prefer if we could wrap every row in its own element and apply grid on that, and also apply grid on the element containing the rows.

It would also be better to avoid inline styling and instead use a CSS variable, I know they are currently unused in our codebase, but I feel like that's not reason enough for us not to.

Basically having: `style={"--columns": columns.length}` and have the actual CSS in the proper files.

Reply:
The issue is that then columns don't scale correctly as each row element is unaware of the next, and could result in mismatched column sizes.

I'd love to be able to do this, but until [CSS subgrid is supported in browsers](https://caniuse.com/css-subgrid), it's basically impossible.

RE CSS custom properties: I was under the impression that CSS custom properties had far worse support than they actually do (~95%). Let's use them!

Reply:
Here's an example of what it looks like if we have cells in row elements.

![image](https://user-images.githubusercontent.com/7406822/112193641-ffd4f580-8bff-11eb-88a9-67d089895043.png)

Reply:
Yeah, I didn't think of that, we would be obliged to use fixed widths to make it happen, which isn't flexible, it's a shame subgrid would take years before it can be applied. Let's keep it as is for now then.

Btw what about accessibility ? At a first glance it seems suggested to use actual tables for presenting data, but I'm not well versed in this to know. I'm gonna research this later.